### PR TITLE
contracts: Deprecate random interface

### PR DESCRIPTION
**Pull request by athei**
This PR deprecates the random API provided by `pallet-contracts`. This means no new codes nor contracts can be deployed using this API. The functionality was already removed from ink! and solang:
https://github.com/paritytech/ink/pull/1442
https://github.com/hyperledger/solang/pull/1041

# Motivation

Randomess shouldn't be part of the core API of `pallet-contracts`.  Reason is that not every chain might be able to provide secure (unpredictable) random numbers. By having it in the core API would force them to either provide an insecure implementation or not use `pallet-contracts` at all. Instead, chains that can provide secure randomness should expose this as a chain extension or via a oracle contract.

# Details

This introduces a new attribute `#[deprecated]` attribute for our `define_env` macro. It makes the annotated API in the following way:
https://github.com/paritytech/substrate/blob/9981e1fc8e4615d77143d876b149c179fa67e28a/frame/contracts/src/lib.rs#L219-L227

It is important to understand that deprecated is different from just being replaced by a new version. If there is a new API for a new version we do not deprecate an API necessarily. The old versions usually stay fully functional. Deprecated on the other hand has the implications mentioned above. We only do that if we plan to eventually drop the API in question. This is the case if the  API:

1. is considered unsafe (as in this case)
2. presents a big maintenance burden
3. is not used anyone anymore
4.  is rendered nonsensical by changes to the logic of the pallet

This is why we also removed all other mentions of the word `deprecated` in other contexts like when a new version is available.

In addition to the random API we also deprecate functions related to the no longer present state rent (replaced by state deposit). This deprecation fulfils point **4** and probably **3**.

### CODE REVIEW

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2299257

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2299354

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2299400

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2299402

File: frame/contracts/proc-macro/src/lib.rs [hunk 1]
```diff
@@ -162,6 +162,8 @@ struct HostFn {
 	returns: HostFnReturn,
 	is_stable: bool,
 	alias_to: Option<String>,
+	/// Formulating the predicate inverted makes the expression using it simpler.
+	is_undeprecated: bool,
```
Comment (xermicus) [inline frame/contracts/proc-macro/src/lib.rs:166]: I don't think that `undeprecated` exists as a word. How about `not_deprecated`?

**Review by xermicus — APPROVED** [review]

File: frame/contracts/proc-macro/src/lib.rs (hunk 1 above)
Comment (athei) [reply frame/contracts/proc-macro/src/lib.rs]: Yes. Good idea :)

**Review by agryaznov — COMMENTED** [review]
How about marking with `#[deprecated]` both versions of `set_rent_allowance()` ?

File: frame/contracts/src/wasm/mod.rs [hunk 2]
```diff
@@ -220,7 +223,16 @@ impl<T: Config> PrefabWasmModule<T> {
 		let module = Module::new(&engine, code)?;
 		let mut store = Store::new(&engine, host_state);
 		let mut linker = Linker::new();
-		E::define(&mut store, &mut linker, T::UnsafeUnstableInterface::get())?;
+		E::define(
+			&mut store,
+			&mut linker,
+			if T::UnsafeUnstableInterface::get() {
```
[the remark this answers is not in the document]
Comment (athei) [reply frame/contracts/src/wasm/mod.rs:229]: The whole point of these enums to see what these arguments are about at the call site. Passing bools into functions just make it hard to read.

**Comment by athei** [comment]
> How about marking with `#[deprecated]` both versions of `set_rent_allowance()` ?

Yes! I missed this one. Deprecated it. I also regrouped so all the deprecated storage rent functions are next to each other.

**Review by agryaznov — APPROVED** [review]

**Comment by athei** [comment]
bot merge

**Comment by Polkadot-Forum** [comment]
This pull request has been mentioned on **Polkadot Forum**. There might be relevant details there:

https://forum.polkadot.network/t/substrate-developer-newsletter-february-2023-notable-updates-for-frame-devs-new-deep-dives-released-calling-builders-to-apply-to-rfps-and-bounties/2137/1

**Pull request merged**